### PR TITLE
[vulkan] Extract GUI utility functions out in samples (NFC)

### PR DESCRIPTION
**Pull request by antiagainst**
These utility functions can be shared for other GUI applications.
This commit is purely shuffling code around, so NFC.

### CODE REVIEW

**Comment by antiagainst** [comment]
@ScottTodd, I've extracted the utility functions out as we discussed last time. Let me know whether this is something you expected. :)

File: iree/tools/vulkan_gui_util.h [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+// Returns the names of the Vulkan instance extensions needed for the given IREE
+// |vulkan_features|.
+std::vector<const char*> GetInstanceExtensions(
+    SDL_Window* window, iree_hal_vulkan_features_t vulkan_features);
+
+// Initializes the Vulkan environment with the given |vulkan_features| and
+// layers/extensions, and writes various Vulkan handles.
+//
+// This function creates Vulkan |instance|, selects a GPU and
+// |lqueue_family_index| with both graphics and compute bits, gets the
```
Comment (ScottTodd) [inline iree/tools/vulkan_gui_util.h:50]:
```suggestion
// |queue_family_index| with both graphics and compute bits, gets the
```

typo

File: iree/tools/vulkan_gui_util.h [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+std::vector<const char*> GetInstanceLayers(
+    iree_hal_vulkan_features_t vulkan_features);
+
+// Returns the names of the Vulkan instance extensions needed for the given IREE
+// |vulkan_features|.
+std::vector<const char*> GetInstanceExtensions(
+    SDL_Window* window, iree_hal_vulkan_features_t vulkan_features);
+
+// Initializes the Vulkan environment with the given |vulkan_features| and
+// layers/extensions, and writes various Vulkan handles.
```
Comment (ScottTodd) [inline iree/tools/vulkan_gui_util.h:47]: How are failures handled? Assert fail, or return nullptrs? (Assert is fine for sample code, good to document it here though)

File: iree/tools/vulkan_gui_util.h [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+                 const VkAllocationCallbacks* allocator, VkInstance* instance,
+                 uint32_t* queue_family_index,
+                 VkPhysicalDevice* physical_device, VkQueue* queue,
+                 VkDevice* device, VkDescriptorPool* descriptor_pool);
+
+// Sets up a ImGui Vukan GUI window.
+//
+// This function creates surface, swapchain, framebuffer, and others in
+// prepration for rendering.
+void SetupVulkanWindow(ImGui_ImplVulkanH_Window* wd,
```
Comment (ScottTodd) [inline iree/tools/vulkan_gui_util.h:66]: I'm on the fence about sticking to C style coding or pulling in C++ concepts like classes, but we could abstract most of these details into a class for samples/tools to hook into

```
class VulkanSample {
  public:
    void Init();
    void RunUntilFinished();
    ...
  protected:
    void Update();
    void Draw();

    VkDevice device_;
    ...
  private:
    void SetupVulkanWindow();
    ...
}
```

Partly depends on how similar the other samples/tools look and if they use ImGui/SDL like the current sample does. This refactor LGTM as a start and we can see if refactoring further makes sense later on.

File: iree/tools/vulkan_gui_util.h [hunk 4]
```diff
@@ -0,0 +1,83 @@
+// Copyright 2020 Google LLC
```
Comment (benvanik) [inline iree/tools/vulkan_gui_util.h:1]: doesn't really seem right for this to be under tools/ (directory is getting crowded) - what is the intent of this? can it be moved under iree/testing/vulkan/ to denote that it's purely test related?

Comment (ScottTodd) [reply iree/tools/vulkan_gui_util.h:1]: `iree/testing/vulkan` would also work. The intent was to share between the Vulkan GUI sample and a new GUI program like `iree-run-module` for running models through vendor-specific tools like NVIDIA Nsight.

Comment (antiagainst) [reply iree/tools/vulkan_gui_util.h:1]: Yeah `iree/testing/vulkan` works for me. Another option might be `iree/tools/utils`.

File: iree/tools/vulkan_gui_util.h (hunk 3 above)
Comment (antiagainst) [reply iree/tools/vulkan_gui_util.h:66]: Yeah, it might be good for internally sharing some of the Vulkan handles. But it can happen next.

**Comment by antiagainst** [comment]
Tested on Windows the `vulkan_inference_gui` app still works. Linux is failing on main due to some other issues that can be fixed separately.

**Review by ScottTodd — APPROVED** [review]

**Pull request merged**